google/cloud/storage/internal/async/rewriter_connection_tracing_test.cc

```diff
@@ -0,0 +1,147 @@
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+#include "google/cloud/storage/internal/async/rewriter_connection_tracing.h"
+#include "google/cloud/storage/mocks/mock_async_rewriter_connection.h"
+#include "google/cloud/storage/testing/canonical_errors.h"
+#include "google/cloud/internal/opentelemetry.h"
+#include "google/cloud/opentelemetry_options.h"
+#include "google/cloud/testing_util/async_sequencer.h"
+#include "google/cloud/testing_util/opentelemetry_matchers.h"
+#include "google/cloud/testing_util/status_matchers.h"
+#include <gmock/gmock.h>
+#include <opentelemetry/trace/semantic_conventions.h>
+#include <cstdint>
+
+namespace google {
+namespace cloud {
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+using ::google::cloud::storage_experimental::RewriteObjectResponse;
+using ::google::cloud::storage_mocks::MockAsyncRewriterConnection;
+using ::google::cloud::testing_util::AsyncSequencer;
+using ::google::cloud::testing_util::EventNamed;
+using ::google::cloud::testing_util::InstallSpanCatcher;
+using ::google::cloud::testing_util::OTelAttribute;
+using ::google::cloud::testing_util::SpanEventAttributesAre;
+using ::google::cloud::testing_util::SpanHasEvents;
+using ::google::cloud::testing_util::SpanHasInstrumentationScope;
+using ::google::cloud::testing_util::SpanKindIsClient;
+using ::google::cloud::testing_util::SpanNamed;
+using ::google::cloud::testing_util::SpanWithStatus;
+using ::google::cloud::testing_util::StatusIs;
+using ::google::cloud::testing_util::ThereIsAnActiveSpan;
+using ::testing::IsEmpty;
+
+auto PermanentError() {
+  return StatusOr<RewriteObjectResponse>(
+      storage::testing::canonical_errors::PermanentError());
+}
+
+TEST(RewriterTracingConnection, Basic) {
+  auto span_catcher = InstallSpanCatcher();
+  AsyncSequencer<void> sequencer;
+
+  auto mock = std::make_unique<MockAsyncRewriterConnection>();
+  EXPECT_CALL(*mock, Iterate)
+      .WillOnce([&] {
+        EXPECT_TRUE(ThereIsAnActiveSpan());
+        return sequencer.PushBack("Iterate(1)").then([](auto) {
+          return PermanentError();
+        });
+      })
+      .WillOnce([&] {
+        EXPECT_TRUE(ThereIsAnActiveSpan());
+        return sequencer.PushBack("Iterate(2)").then([](auto) {
+          return make_status_or(
+              RewriteObjectResponse{1000, 3000, "test-token", absl::nullopt});
+        });
+      })
+      .WillOnce([&] {
+        EXPECT_TRUE(ThereIsAnActiveSpan());
+        return sequencer.PushBack("Iterate(3)").then([](auto) {
+          return make_status_or(
```

Comment:
I think we want to verify the span is not active in these continuation:

```cc
        return sequencer.PushBack("Iterate(3)").then([](auto) {
          EXPECT_FALSE(ThereIsAnActiveSpan());
          EXPECT_FALSE(OTelContextCaptured());
          return ...;
        });
```

e.g. https://github.com/googleapis/google-cloud-cpp/blob/5a1fe557ae3c37e0611322099dde29e4bfd0d730/google/cloud/storage/internal/async/connection_tracing_test.cc#L96-L97

Reply:
Done